eos-reclaim-swap

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python3
+import argparse
+import os
+import parted
+import subprocess
+
+class SwapRemover():
+    '''
+    Remove swap partitions from a disk and reclaim their space.
+    '''
+
+    def get_table_str(self, disk):
+        '''
+        Return a text string with the disk partition table.
+        '''
+        cmd = ['fdisk', '-l', disk]
+        out = subprocess.run(cmd, stdout=subprocess.PIPE).stdout
+        return out.decode()
+
+
+    def find_disk(self, partition):
+        '''
+        Return path to a block device representing the disk containing
+        partition.
+        '''
+        p = os.path.realpath(partition)
+        disk = None
+
+        # /dev/sd[a-z][0-9]
+        if (p[:-2] == '/dev/sd'):
+            disk = p[:-1]
+            self.sep = ''
+
+        # /dev/[a-z]+p[0-9], but not /dev/sdp[0-9] (eMMC)
+        elif (p.startswith('/dev/') and p[-2:-1] == 'p'):
+            disk = p[:-2]
+            self.sep = 'p'
+
+        # device-mapped images (dual-boot setups)
+        if (p[-1] == '/dev/mapper/endless-image'):
+            disk = partition[:-1]
+            self.sep = ''
+
+        # for testing with loop devices
+        if (p[:-3] == '/dev/loop'):
+            self.sep = 'p'
+
+        # for testing with image files
+        if not p.startswith('/dev/'):
+            disk = partition[:-1]
+            self.sep = ''
+
+        return disk
+
+
+    def reclaim_swap_space(self, disk_path):
+        '''
+        Remove the first swap partition on disk and expand the preceding
+        partition so it spawns over the space previously used by the swap
+        partition.
+        '''
+        dev = parted.getDevice(disk_path)
+        disk = parted.newDisk(dev)
+        parts = disk.partitions
+
+        for p in range(len(parts)):
+            # (py)parted does not expose the MBR partition ID or the GPT GUID,
+            # so we need to use the file system signature to identify swap
+            # partitions and thus this only works with formatted swap
+            # partitions. When the fileSystem property is set on swap
+            # partitions it can be either linux-swap(v0) or linux-swap(v1).
+            if parts[p].fileSystem and parts[p].fileSystem.type[:-4] == 'linux-swap':
+                geom = parted.Geometry(start=parts[p-1].geometry.start,
+                                       end=parts[p].geometry.end, device=dev)
+                constraint = parted.Constraint(maxGeom=geom)
+                disk.removePartition(parts[p])
+                disk.maximizePartition(parts[p-1], constraint)
```

Comment:
This implicitly assumes that the logical partition order matches the on-disk order. Probably safe enough but it would be interesting to test what happens if it doesn't.

Reply:
Yes, because that is how all our layouts are laid-out. If for some reason `p==0` here, `parted.Geometry()` will raise an exception because start > end, which I believe is the correct way of failing here since this is not an expected input. We could check for `p >= 1` and silently do nothing, but that would be masking a potentially bigger problem (why the heck we have swap as the first partition?).